File: python/centroid.py
import json
from collections import defaultdict

class Config:
    def __init__(self, config):
        if type(config) is dict or type(config) is list:
            self.raw_config = config
        else:
            self.raw_config = json.loads(config)
            self._validate_unique_keys()

    def __getattr__(self, attrib):
        return self[attrib]

    # config['key']
    def __getitem__(self, key):
        value = _get_value(key, self.raw_config)
        if type(value) is dict or type(value) is list:
            return Config(value)
        return value

    def __iter__(self):
        if type(self.raw_config) is dict:
            for key, value in self.raw_config.iteritems():
                if type(value) is dict:
                    yield key, Config(value)
                else:
                    yield key, value

        else:
            for value in self.raw_config:
                if type(value) is dict:
                    yield Config(value)
                else:
                    yield value

    # to string
    def __str__(self):
        return str(json.dumps(self.raw_config))

    def _validate_unique_keys(self):
        normalized_keys = defaultdict(list)
        for key, _ in self.raw_config.iteritems():
            normalized_keys[_get_normalised_key(key)].append(key)

        dups = list()
        for _, value in normalized_keys.iteritems():
            if len(value) > 1:
                dups.extend(value)

        if len(dups) > 0:
            raise KeyError("centroid.Config instance contains duplicate keys: " + ", ".join(dups))

    def for_environment(self, env):
        env_json = self.raw_config[env]
        env_json["environment"] = env

        actual_key = _get_actual_key('all', self.raw_config)
        if actual_key is None:
            return Config(env_json)

        all_json = _get_value(actual_key, self.raw_config)

        return Config(_dict_merge(all_json, env_json))

    def __contains__(self, key):
        return _get_actual_key(key, self.raw_config) is not None

    @staticmethod
    def from_file(filename):
        with open(filename) as json_file:
            str_json = json_file.read()
            return Config(str_json)

# case insensitive hashtable helpers
def _get_normalised_key(unnormalisedKey):
    return unnormalisedKey.replace('_','').lower()

def _get_value(key, hashtable):
    if type(key) is int:
        return hashtable[key]

    actual_key = _get_actual_key(key, hashtable)
    if actual_key is None:
        raise KeyError('centroid.Config instance does not contain key: ' + key)
    return hashtable[actual_key]

def _get_actual_key(key, hashtable):
    result = [ k for k in hashtable.keys() if _get_normalised_key(key) == _get_normalised_key(k) ]
    if len(result) > 0:
        return result[0]
    return None

def _dict_merge(left, right):
    if not isinstance(right, dict):
        return right
    for k, v in right.iteritems():
        if k in left and isinstance(left[k], dict):
            left[k] = _dict_merge(left[k], v)
        else:
            left[k] = v
    return left

File: ruby/lib/centroid.rb
require "json"

module Centroid
  class Config
    attr_reader :raw_config

    def initialize(config)
      if config.is_a?(Hash)
        @raw_config = config
      else
        @raw_config = JSON.parse(config)
        validate_unique_keys!
      end
    end

    def method_missing(attrib, *)
      self[attrib]
    end

    def respond_to_missing?(method_name, include_all)
      has_key?(method_name)
    end

    def each
      return enum_for :each unless block_given?

      raw_config.each do |key, value|
        if value.is_a?(Hash)
          yield key, Config.new(value)
        else
          yield key, value
        end
      end
    end

    def [](key)
      raise KeyError.new("Centroid::Config instance does not contain key: #{key}") unless has_key?(key)

      value = find_value(key)

      if value.is_a?(Hash)
        Config.new(value)
      elsif value.is_a?(Array)
        value.map { |e| e.is_a?(Hash) ? Config.new(e) : e }
      else
        value
      end
    end

    def has_key?(key)
      !!actual_key(key)
    end

    alias_method :include?, :has_key?

    def to_s
      JSON.dump(raw_config)
    end

    def for_environment(env)
      env_json = raw_config[env]
      all_key = actual_key("all")
      if all_key.nil?
        config = Config.new(env_json)
      else
        all_json = raw_config[all_key]
        config = Config.new(deep_merge(all_json, env_json))
      end

      if !config.has_key?('environment')
        config.raw_config['environment'] = env
      end

      config
    end

    def self.from_file(filename)
      str_json = File.read(filename)
      self.new(str_json)
    end

    private

    def normalize_key(unnormalised_key)
      unnormalised_key.to_s.gsub("_", "").downcase
    end

    def find_value(key)
      raw_config[actual_key(key)]
    end

    def actual_key(key)
      raw_config.keys.find { |k| normalize_key(key) == normalize_key(k) }
    end

    def validate_unique_keys!
      normalized_keys = raw_config.keys.map { |k| { key: k, normalized_key: normalize_key(k) } }
      dups = normalized_keys.group_by { |nk| nk[:normalized_key] }.select { |_, g| g.size > 1 }

      return if dups.empty?

      keys = dups.values.flat_map { |d| d.map { |e| e[:key] } }
      raise KeyError, "Centroid::Config instance contains duplicate keys: #{keys.join(', ')}"
    end

    def deep_merge(left, right)
      return right if not right.is_a?(Hash)

      right.each_pair do |k, rv|
        lv = left[k]
        left[k] = if lv.is_a?(Hash) && rv.is_a?(Hash)
          deep_merge(lv, rv)
        else
          rv
        end
      end

      left
    end
  end
end
